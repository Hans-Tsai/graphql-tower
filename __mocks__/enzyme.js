import { configure, ShallowWrapper } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

// https://github.com/airbnb/enzyme/issues/1295
ShallowWrapper.prototype.diveTo = function diveTo(name, options) {
  let component = this;

  while (component.name() !== name) {
    component = component.dive(options);
  }
  component = component.dive();

  return component;
};

export * from 'enzyme';
