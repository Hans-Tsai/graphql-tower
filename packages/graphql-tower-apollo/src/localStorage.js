import global from 'global/window';

export default global.localStorage || {
  setItem: () => undefined,
  getItem: () => undefined,
};
