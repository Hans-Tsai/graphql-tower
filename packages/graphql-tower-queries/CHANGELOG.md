# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="3.29.0"></a>
# [3.29.0](https://github.com/tmotx/graphql-tower/compare/v3.28.9...v3.29.0) (2017-12-23)


### Features

* **queries:** support totalCount on Connection ([aa9bc56](https://github.com/tmotx/graphql-tower/commit/aa9bc56))




<a name="3.26.0"></a>
# [3.26.0](https://github.com/tmotx/graphql-tower/compare/v3.25.2...v3.26.0) (2017-12-13)




**Note:** Version bump only for package graphql-tower-queries

<a name="3.24.2"></a>
## [3.24.2](https://github.com/tmotx/graphql-tower/compare/v3.24.1...v3.24.2) (2017-12-11)


### Bug Fixes

* **queries:** support data and contextValue on payload ([8cd7ef3](https://github.com/tmotx/graphql-tower/commit/8cd7ef3))




<a name="3.24.1"></a>
## [3.24.1](https://github.com/tmotx/graphql-tower/compare/v3.24.0...v3.24.1) (2017-12-10)


### Bug Fixes

* **redis-subscriptions:** added onMessage to support each params ([53d8c45](https://github.com/tmotx/graphql-tower/commit/53d8c45))




<a name="3.21.3"></a>
## [3.21.3](https://github.com/tmotx/graphql-tower/compare/v3.21.2...v3.21.3) (2017-12-05)




**Note:** Version bump only for package graphql-tower-queries

<a name="3.21.2"></a>
## [3.21.2](https://github.com/tmotx/graphql-tower/compare/v3.21.1...v3.21.2) (2017-12-01)


### Bug Fixes

* **queries:** run middleware on withFilter ([5e0a20b](https://github.com/tmotx/graphql-tower/commit/5e0a20b))




<a name="3.21.0"></a>
# [3.21.0](https://github.com/tmotx/graphql-tower/compare/v3.20.1...v3.21.0) (2017-11-30)


### Features

* **queries:** check offset from results ([01efee3](https://github.com/tmotx/graphql-tower/commit/01efee3))




<a name="3.20.0"></a>
# [3.20.0](https://github.com/tmotx/graphql-tower/compare/v3.19.10...v3.20.0) (2017-11-29)


### Features

* **queries:** added subscription ([b7498e0](https://github.com/tmotx/graphql-tower/commit/b7498e0))




<a name="3.12.0"></a>
# [3.12.0](https://github.com/tmotx/graphql-tower/compare/v3.11.0...v3.12.0) (2017-11-14)


### Features

* **queries:** support class name ([b515179](https://github.com/tmotx/graphql-tower/commit/b515179))




<a name="3.1.0"></a>
# [3.1.0](https://github.com/tmotx/graphql-tower/compare/v3.0.1...v3.1.0) (2017-10-26)


### Features

* **package:** setup lerna-semantic-release ([a4fb7d8](https://github.com/tmotx/graphql-tower/commit/a4fb7d8))
