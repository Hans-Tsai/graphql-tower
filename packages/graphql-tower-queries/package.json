{
  "name": "graphql-tower-queries",
  "version": "3.29.0",
  "description": "",
  "main": "lib/index.js",
  "scripts": {
    "build": "rm -rf lib/* && babel src --ignore __tests__ --out-dir lib"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/tmotx/graphql-tower.git"
  },
  "author": "tmotx",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/tmotx/graphql-tower/issues"
  },
  "homepage": "https://github.com/tmotx/graphql-tower#readme",
  "files": [
    "lib"
  ],
  "dependencies": {
    "graphql-tower-errors": "^3.21.3",
    "iterall": "^1.1.3",
    "lodash": "^4.17.4"
  },
  "peerDependencies": {
    "graphql": "^0.11.7"
  }
}
