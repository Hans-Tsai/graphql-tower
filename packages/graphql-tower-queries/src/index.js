export { default as GraphQLField } from './GraphQLField';
export { default as PayloadField } from './PayloadField';
export { default as Query } from './Query';
export { default as QueryWithNode } from './QueryWithNode';
export { default as QueryWithConnection } from './QueryWithConnection';
export { default as Mutation } from './Mutation';
export { default as Subscription } from './Subscription';
export * from './afterware';
export * from './middleware';
