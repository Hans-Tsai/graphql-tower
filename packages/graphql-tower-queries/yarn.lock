# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


async-limiter@~1.0.0:
  version "1.0.0"
  resolved "https://registry.yarnpkg.com/async-limiter/-/async-limiter-1.0.0.tgz#78faed8c3d074ab81f22b4e985d79e8738f720f8"

backo2@^1.0.2:
  version "1.0.2"
  resolved "https://registry.yarnpkg.com/backo2/-/backo2-1.0.2.tgz#31ab1ac8b129363463e35b3ebb69f4dfcfba7947"

es6-promise@^4.1.1:
  version "4.1.1"
  resolved "https://registry.yarnpkg.com/es6-promise/-/es6-promise-4.1.1.tgz#8811e90915d9a0dba36274f0b242dbda78f9c92a"

eventemitter3@^2.0.3:
  version "2.0.3"
  resolved "https://registry.yarnpkg.com/eventemitter3/-/eventemitter3-2.0.3.tgz#b5e1079b59fb5e1ba2771c0a993be060a58c99ba"

graphql-subscriptions@^0.5.5:
  version "0.5.5"
  resolved "https://registry.yarnpkg.com/graphql-subscriptions/-/graphql-subscriptions-0.5.5.tgz#f8ad3e094a23128d5065725f9b60020d3293999d"
  dependencies:
    es6-promise "^4.1.1"
    iterall "^1.1.3"

graphql-tower-errors@^3.16.0:
  version "3.16.0"
  resolved "https://registry.yarnpkg.com/graphql-tower-errors/-/graphql-tower-errors-3.16.0.tgz#5dc7b845ef15dc9eb63fa4401f2c324f0cbc8b91"

iterall@^1.1.1, iterall@^1.1.3:
  version "1.1.3"
  resolved "https://registry.yarnpkg.com/iterall/-/iterall-1.1.3.tgz#1cbbff96204056dde6656e2ed2e2226d0e6d72c9"

lodash.assign@^4.2.0:
  version "4.2.0"
  resolved "https://registry.yarnpkg.com/lodash.assign/-/lodash.assign-4.2.0.tgz#0d99f3ccd7a6d261d19bdaeb9245005d285808e7"

lodash.isobject@^3.0.2:
  version "3.0.2"
  resolved "https://registry.yarnpkg.com/lodash.isobject/-/lodash.isobject-3.0.2.tgz#3c8fb8d5b5bf4bf90ae06e14f2a530a4ed935e1d"

lodash.isstring@^4.0.1:
  version "4.0.1"
  resolved "https://registry.yarnpkg.com/lodash.isstring/-/lodash.isstring-4.0.1.tgz#d527dfb5456eca7cc9bb95d5daeaf88ba54a5451"

lodash@^4.17.4:
  version "4.17.4"
  resolved "https://registry.yarnpkg.com/lodash/-/lodash-4.17.4.tgz#78203a4d1c328ae1d86dca6460e369b57f4055ae"

safe-buffer@~5.1.0:
  version "5.1.1"
  resolved "https://registry.yarnpkg.com/safe-buffer/-/safe-buffer-5.1.1.tgz#893312af69b2123def71f57889001671eeb2c853"

subscriptions-transport-ws@^0.9.1:
  version "0.9.1"
  resolved "https://registry.yarnpkg.com/subscriptions-transport-ws/-/subscriptions-transport-ws-0.9.1.tgz#66488bd5997b6081277a25d6a3015a5aceffed63"
  dependencies:
    backo2 "^1.0.2"
    eventemitter3 "^2.0.3"
    iterall "^1.1.1"
    lodash.assign "^4.2.0"
    lodash.isobject "^3.0.2"
    lodash.isstring "^4.0.1"
    symbol-observable "^1.0.4"
    ws "^3.0.0"

symbol-observable@^1.0.4:
  version "1.0.4"
  resolved "https://registry.yarnpkg.com/symbol-observable/-/symbol-observable-1.0.4.tgz#29bf615d4aa7121bdd898b22d4b3f9bc4e2aa03d"

ultron@~1.1.0:
  version "1.1.1"
  resolved "https://registry.yarnpkg.com/ultron/-/ultron-1.1.1.tgz#9fe1536a10a664a65266a1e3ccf85fd36302bc9c"

ws@^3.0.0, ws@^3.3.2:
  version "3.3.2"
  resolved "https://registry.yarnpkg.com/ws/-/ws-3.3.2.tgz#96c1d08b3fefda1d5c1e33700d3bfaa9be2d5608"
  dependencies:
    async-limiter "~1.0.0"
    safe-buffer "~5.1.0"
    ultron "~1.1.0"
