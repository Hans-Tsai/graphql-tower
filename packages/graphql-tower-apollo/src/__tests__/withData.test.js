/**
 * @jest-environment jsdom
 */
import _ from 'lodash';
import React from 'react';
import http from 'http';
import gql from 'graphql-tag';
import express from 'express';
import ws from 'ws';
import bodyParser from 'body-parser';
import { subscribe, GraphQLID, GraphQLString, GraphQLSchema, GraphQLObjectType } from 'graphql';
import { PubSub } from 'graphql-subscriptions';
import { SubscriptionServer } from 'subscriptions-transport-ws';
import { mount } from 'enzyme';
import { graphql } from 'react-apollo';
import { execute } from 'apollo-link';
import { graphqlExpress } from 'apollo-server-express';
import { Query } from 'graphql-tower-queries';
import { withData, initApollo } from '../';
import localStorage from '../localStorage';

jest.unmock('node-fetch');

const pubsub = new PubSub();

const renewToken = jest.fn();
const resolveSpy = jest.fn(payload => payload * 2);
const subscribeSpy = jest.fn(() => pubsub.asyncIterator('me'));

const User = new GraphQLObjectType({
  name: 'User',
  fields: {
    id: { type: GraphQLID },
    name: { type: GraphQLString },
  },
});

const Me = class extends Query {
  type = User;
  resolve = resolveSpy;
  subscribe = subscribeSpy;
};

const schema = new GraphQLSchema({
  query: new GraphQLObjectType({ name: 'Query', fields: { me: new Me() } }),
  subscription: new GraphQLObjectType({ name: 'Subscription', fields: { me: new Me() } }),
});

const app = express();
app.use('/graphql', bodyParser.json(), graphqlExpress((req, res) => {
  const token = renewToken();
  if (token) {
    res.append('X-Refresh-Token', token);
    res.append('Authorization', token);
  }

  return { schema, context: { authorization: req.headers.authorization } };
}));

const server = http.createServer(app);
server.listen();
const { port } = server.address();

new SubscriptionServer( // eslint-disable-line
  {
    schema, subscribe, execute, onConnect: params => (params : {}),
  },
  { server, path: '/' },
);

class App extends React.Component {
  static async getInitialProps() {
    return { value: 99 };
  }

  render() {
    return (<div />);
  }
}

const apollo = {
  httpUri: `http://localhost:${port}/graphql`,
  wsUri: `ws://localhost:${port}/`,
  webSocketImpl: ws,
};

describe('withData', () => {
  describe('process.browser is false', () => {
    beforeEach(() => { process.browser = false; });

    it('getInitialProps without cookie', async () => {
      resolveSpy.mockReturnValueOnce(Promise.resolve({ id: '50', name: 'hello' }));
      const WrapApp = withData(apollo)(graphql(gql`query { me { id name } }`)(App));

      const props = await WrapApp.getInitialProps();
      expect(props).toMatchSnapshot();

      expect(resolveSpy)
        .toHaveBeenLastCalledWith(undefined, {}, { authorization: undefined }, expect.anything());
      expect(resolveSpy).toHaveBeenCalledTimes(1);

      const component = mount(<WrapApp {...props} />);
      expect(component).toMatchSnapshot();
    });

    it('getInitialProps with cookie', async () => {
      resolveSpy.mockReturnValueOnce(Promise.resolve({ id: '50', name: 'hello' }));
      const WrapApp = withData(apollo)(graphql(gql`query { me { id name } }`)(App));

      const props = await WrapApp.getInitialProps(_.set({}, ['req', 'headers', 'cookie'], 'access_token=key of token'));
      expect(props).toMatchSnapshot();
      expect(resolveSpy)
        .toHaveBeenLastCalledWith(undefined, {}, { authorization: 'Bearer key of token' }, expect.anything());
      expect(resolveSpy).toHaveBeenCalledTimes(1);

      const component = mount(<WrapApp {...props} />);
      expect(component).toMatchSnapshot();
    });

    it('when throw error', async () => {
      resolveSpy.mockReturnValueOnce(Promise.resolve(new Error()));
      const WrapApp = withData(apollo)(graphql(gql`query { me { id name } }`)(App));

      const props = await WrapApp.getInitialProps();
      expect(props).toMatchSnapshot();
      expect(resolveSpy)
        .toHaveBeenLastCalledWith(undefined, {}, { authorization: undefined }, expect.anything());
      expect(resolveSpy).toHaveBeenCalledTimes(1);
    });

    it('when no getInitialProps', async () => {
      resolveSpy.mockReturnValueOnce(Promise.resolve({ id: '50', name: 'hello' }));
      const WrapApp = withData(apollo)(graphql(gql`query { me { id name } }`)(() => <div />));
      expect(await WrapApp.getInitialProps()).toMatchSnapshot();
      expect(resolveSpy).toHaveBeenCalledTimes(1);
    });
  });

  describe('process.browser is true', () => {
    beforeEach(() => {
      process.browser = true;
      initApollo.client = null;
    });

    it('getInitialProps', async () => {
      const WrapApp1 = withData(apollo)(graphql(gql`query { me { id name } }`)(App));
      expect(await WrapApp1.getInitialProps()).toMatchSnapshot();

      const WrapApp2 = withData(apollo)(graphql(gql`query { me { id name } }`)(App));
      expect(await WrapApp2.getInitialProps()).toMatchSnapshot();

      expect(resolveSpy).toHaveBeenCalledTimes(0);
    });

    it('has not websocket', async () => {
      const WrapApp = withData({ ...apollo, wsUri: undefined })(graphql(gql`query { me { id name } }`)(App));
      expect(await WrapApp.getInitialProps()).toMatchSnapshot();
      expect(resolveSpy).toHaveBeenCalledTimes(0);
    });
  });

  describe('initApollo', () => {
    beforeEach(() => {
      process.browser = true;
      initApollo.client = null;
    });

    it('init', async () => {
      expect(initApollo()).toMatchSnapshot();
      expect(initApollo()).toBe(initApollo());
    });

    it('subscription', async () => {
      let resolve;
      const promise = new Promise((__) => { resolve = __; });
      subscribeSpy.mockImplementationOnce(resolve);

      const client = initApollo({}, { ...apollo, authorization: 'key of token' });
      execute(client.link, { query: gql`subscription { me { id name } }` }).subscribe({});

      await promise;

      expect(subscribeSpy).toHaveBeenLastCalledWith(undefined, {}, { authorization: 'key of token' }, expect.anything());
      expect(subscribeSpy).toHaveBeenCalledTimes(1);
    });

    it('query', async () => {
      renewToken.mockReturnValueOnce('new token');

      let resolve;
      const promise = new Promise((__) => { resolve = __; });
      resolveSpy.mockImplementationOnce(resolve);

      const client = initApollo({}, { ...apollo, authorization: 'key of token' });
      execute(client.link, { query: gql`query { me { id name } }` }).subscribe({});

      await promise;

      await new Promise(setImmediate);
      await new Promise(setImmediate);

      expect(resolveSpy).toHaveBeenLastCalledWith(undefined, {}, { authorization: 'key of token' }, expect.anything());
      expect(resolveSpy).toHaveBeenCalledTimes(1);
    });

    it('authorization', async () => {
      localStorage.setItem = jest.fn();
      renewToken.mockReturnValueOnce('new token');

      const client = initApollo({}, apollo);

      let resolve1;
      const promise1 = new Promise((__) => { resolve1 = __; });
      resolveSpy.mockImplementationOnce(resolve1);
      execute(client.link, { query: gql`query { me { id name } }` }).subscribe({});

      await promise1;

      await new Promise(setImmediate);
      await new Promise(setImmediate);

      let resolve2;
      const promise2 = new Promise((__) => { resolve2 = __; });
      resolveSpy.mockImplementationOnce(resolve2);
      execute(client.link, { query: gql`query { me { id name } }` }).subscribe({});

      await promise2;

      await new Promise(setImmediate);
      await new Promise(setImmediate);

      expect(resolveSpy)
        .toHaveBeenCalledWith(undefined, {}, { authorization: undefined }, expect.anything());
      expect(resolveSpy)
        .toHaveBeenCalledWith(undefined, {}, { authorization: 'Bearer new token' }, expect.anything());
      expect(resolveSpy).toHaveBeenCalledTimes(2);
      expect(localStorage.setItem).toHaveBeenLastCalledWith('graphql-tower-token', 'new token');
      expect(localStorage.setItem).toHaveBeenCalledTimes(1);
    });
  });
});
