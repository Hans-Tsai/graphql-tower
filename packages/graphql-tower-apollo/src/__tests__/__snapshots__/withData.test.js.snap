// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withData initApollo init 1`] = `
ApolloClient {
  "cache": InMemoryCache {
    "addTypename": true,
    "config": Object {
      "addTypename": true,
      "cacheResolvers": undefined,
      "dataIdFromObject": undefined,
      "fragmentMatcher": IntrospectionFragmentMatcher {
        "isReady": false,
        "match": [Function],
      },
      "storeFactory": [Function],
    },
    "data": ObjectCache {
      "data": Object {},
    },
    "optimistic": Array [],
    "silenceBroadcast": false,
    "watches": Array [],
  },
  "defaultOptions": Object {},
  "disableNetworkFetches": false,
  "link": ApolloLink {
    "request": [Function],
  },
  "mutate": [Function],
  "query": [Function],
  "queryDeduplication": true,
  "reFetchObservableQueries": [Function],
  "resetStore": [Function],
  "ssrMode": false,
  "store": DataStore {
    "cache": InMemoryCache {
      "addTypename": true,
      "config": Object {
        "addTypename": true,
        "cacheResolvers": undefined,
        "dataIdFromObject": undefined,
        "fragmentMatcher": IntrospectionFragmentMatcher {
          "isReady": false,
          "match": [Function],
        },
        "storeFactory": [Function],
      },
      "data": ObjectCache {
        "data": Object {},
      },
      "optimistic": Array [],
      "silenceBroadcast": false,
      "watches": Array [],
    },
  },
  "version": "2.0.4",
  "watchQuery": [Function],
}
`;

exports[`withData process.browser is false getInitialProps with cookie 1`] = `
Object {
  "cache": Object {
    "$ROOT_QUERY.me": Object {
      "__typename": "User",
      "id": "50",
      "name": "hello",
    },
    "ROOT_QUERY": Object {
      "me": Object {
        "generated": true,
        "id": "$ROOT_QUERY.me",
        "type": "id",
      },
    },
  },
  "value": 99,
}
`;

exports[`withData process.browser is false getInitialProps with cookie 2`] = `
<div>
  50-hello value: 99
</div>
`;

exports[`withData process.browser is false getInitialProps without cookie 1`] = `
Object {
  "cache": Object {
    "$ROOT_QUERY.me": Object {
      "__typename": "User",
      "id": "50",
      "name": "hello",
    },
    "ROOT_QUERY": Object {
      "me": Object {
        "generated": true,
        "id": "$ROOT_QUERY.me",
        "type": "id",
      },
    },
  },
  "value": 99,
}
`;

exports[`withData process.browser is false getInitialProps without cookie 2`] = `
<div>
  50-hello value: 99
</div>
`;

exports[`withData process.browser is false when NoSSR 1`] = `
Object {
  "value": 99,
}
`;

exports[`withData process.browser is false when NoSSR 2`] = `
<div>
  - value: 99
</div>
`;

exports[`withData process.browser is false when no getInitialProps 1`] = `
Object {
  "cache": Object {
    "$ROOT_QUERY.me": Object {
      "__typename": "User",
      "id": "50",
      "name": "hello",
    },
    "ROOT_QUERY": Object {
      "me": Object {
        "generated": true,
        "id": "$ROOT_QUERY.me",
        "type": "id",
      },
    },
  },
  "context": undefined,
  "url": Object {},
}
`;

exports[`withData process.browser is false when throw error 1`] = `
Object {
  "value": 99,
}
`;

exports[`withData process.browser is true getInitialProps 1`] = `
Object {
  "value": 99,
}
`;

exports[`withData process.browser is true getInitialProps 2`] = `
Object {
  "value": 99,
}
`;

exports[`withData process.browser is true has not websocket 1`] = `
Object {
  "value": 99,
}
`;
